library/src/main/java/kwasm/format/text/Type.kt

```diff
@@ -14,37 +14,54 @@
 
 package kwasm.format.text
 
+import kwasm.ast.ValueType
 import kwasm.format.ParseContext
+import kwasm.format.ParseException
 
-sealed class Type(
+sealed class Type<T>(
     protected val sequence: CharSequence,
     protected val context: ParseContext? = null
 ) {
+    val value: T by lazy {
+        parseValue()
+    }
 
-    protected abstract fun parseValue(): Unit
+    protected abstract fun parseValue(): T
 
     class ValueType(
         sequence: CharSequence,
         context: ParseContext? = null
-    ) : Type(sequence, context) {
-        override fun parseValue() {
-            TODO("not implemented")
+    ) : Type<kwasm.ast.ValueType>(sequence, context) {
+        override fun parseValue(): kwasm.ast.ValueType {
+            return when (sequence){
+                "i32" -> kwasm.ast.ValueType.I32
+                "i64" -> kwasm.ast.ValueType.I64
+                "f32" -> kwasm.ast.ValueType.F32
+                "f64" -> kwasm.ast.ValueType.F64
+                else -> {
+                    throw ParseException("Invalid ValueType", context)
+                }
+            }
         }
     }
 
     class ResultType(
         sequence: CharSequence,
         context: ParseContext? = null
-    ) : Type(sequence, context) {
-        override fun parseValue() {
-            TODO("not implemented")
+    ) : Type<kwasm.ast.ValueType?>(sequence, context) {
+        override fun parseValue(): kwasm.ast.ValueType? {
+            val keywordAndParameters = getOperationAndParameters(sequence, context)
+            if(keywordAndParameters.first != "result" || keywordAndParameters.second.size > 1){
+                return null
+            }
```

Comment:
This should probably throw a ParseException.

I think we'd only want to return `null` if the `sequence` is empty.

Reply:
So this was a bit tricky for me to think about. When parsing something that may or may not exist, will we ever be passing in ""? So my thought was if the keyword mismatched, the null would let us know that it didn't exist. However, now I think the best plan is your suggestion, and we can just catch the exception when parsing ResultTypes and check what happened.